### ISSUE TITLE
Split mail delivery GSuite <--> Mailu

### ISSUE THREAD

**Issue by deze333** [Mailu/Mailu#1332]
Has anyone had success with **split email delivery** placing Mailu "behind" GSuite (or whatever) mail server?

Split delivery config well described in [this doc from iRedMail](https://gist.github.com/iredmail/fda90a4fd0fc4453f6a2c6c18638e047):

>  If you're hosting same mail domain on 2 servers, e.g. Google G-suite and iRedMail server, you can follow this tutorial to update iRedMail server settings to relay non-existing addresses (under the hosted mail domain) to G-suite. For example, mailbox `xyz@domain.com` exists on Gsuite but NOT on iRedMail, with this tutotial, you're able to relay email sent to `xyz@` to Gsuite.

My simplistic implementation looks like this:

**`main.cf`**

```ini
transport_maps = pcre:/overrides/transport_maps
smtpd_sender_login_maps = pcre:/overrides/sender_logins, ${podop}senderlogin

# Disable sender address matching to allow all, not just known addresses
smtpd_client_restrictions =
  ...
  # check_sender_access ${podop}senderaccess,
  ...
```

**`/overrides/transport_maps`**

```ini
# Relay any user that belongs to our domain
/^.+domain.com$/    smtp:[smtp-relay.gmail.com]:25  
```

**`/overrides/sender_logins`**

```ini
# Allow all users from our domain
 /^(.*)@domain.com$/   ${1}@domain.com
```
This config obviously creates a "too many hops" problem when sending from GSuite address to our mail server address, as our mail server simply relays that email back to GSuite.

Is there a way to tackle this at the config level, or this can only be solved via adding a podop service?

**Comment by kaiyou** [comment]
This is in fact a duplicate of #350. We already have relay enabled for domains, simply using the relay feature.

At the time of the referenced issue it was difficult to implement because everything relied in database queries. Now that we have podop and API calls implemented, it should be doable to add a "fallback MX" field to domains, and use dynamic transport maps for the rest.

The issue with your config is that all email is forwarded back to gmail, not that of non existing local addresses. Implementing this behavior in Postfix directly might prove tricky.

**Comment by kaiyou** [comment]
TL;DR this is tricky to do with postfix directly but should be okay using podop, any contribution to implement it is welcome.

**Comment by deze333** [comment]
Thank you for the explanation! Some more clarification please, if you don't mind.

I've had a look at podop and it appears to be a mechanism for a basic "key-to-value" retrieval. 

Just to make sure we're on the same page, suppose we have some users on `GSuite`, others on `domain.com` but all belonging to the same `domain.com`:

GSuite | domain.com
--- | --- |
a@domain.com | c@domain.com
b@domain.com | d@domain.com

This means that Mailu only knows about users `c` and `d` and therefore should perform an SQL query to figure out what transport to use: `local` for locals or `[smtp-relay.gmail.com]:25` for the rest.

Borrowing from iRedMail example, they use queries like:

**`transport_maps_split.cf`**

```sql
SELECT split_transport FROM domain WHERE domain='%d' AND backupmx=0 AND split_transport<>'' AND active=1
```

**`virtual_mailbox_split.cf`**

```sql
SELECT split_transport FROM domain WHERE NOT EXISTS (SELECT address FROM forwardings WHERE address='%s' AND active=1 LIMIT 1) AND domain='%d' AND backupmx=0 AND split_transport<>'' AND active=1
```
In Mailu case the query logic will be like "return external smtp if address not in user table".

Now I am somewhat confused as to how to use podop to achieve that kind of queries?

**Comment by kaiyou** [comment]
Podop is already configured in Mailu for mapping transports and the current API implementation handles relay domains: https://github.com/Mailu/Mailu/blob/9db709515a45afbbc1045a079c5e64498d631fd7/core/admin/mailu/internal/views/postfix.py#L33

We would need to alter the domain model to add a fallback MX field and use that field in the API function to dynamically return the proper MX if required.

After checking if a relay domain exist, we should check if a domain exists and has a fallback MX configured, then if it has, check if the email resolves before returning the transport if necessary.

**Comment by deze333** [comment]
Aha, thanks for pointing at the code where it all happens, somehow I couldn't locate it. So for now I can add a simple code hack like below while waiting for the schema update.

```python
@internal.route("/postfix/transport/<path:email>")
def postfix_transport(email):
    if email == '*' or re.match("(^|.*@)\[.*\]$", email):
        return flask.abort(404)
    localpart, domain_name = models.Email.resolve_domain(email)
    #  Workaround: If email doesn't resolve, send to relay
    if localpart is None or domain_name is None:
        return flask.jsonify("smtp:[{}]:{}".format("smtp-relay.gmail.com", "25"))
    #  End of workaround
    relay = models.Relay.query.get(domain_name) or flask.abort(404)
    return flask.jsonify("smtp:[{}]".format(relay.smtp))
```

Will try it out over the weekend.

**Comment by kaiyou** [comment]
Not sure about your workaround, I am curious if it behaves like you expect, since I am pretty sure we need to make one more query to the database.

**Comment by deze333** [comment]
You're right, I rushed with the previous code. 

Here's the code that works and no overrides or any changes needed in `main.cf`.
Tested and yes, I am able to exchange emails with Gmail while staying on the same domain, awesome. 

**`Mailu/core/admin/mailu/internal/views/postfix.py`**

```python
@internal.route("/postfix/mailbox/<path:email>")
def postfix_mailbox_map(email):
    user = models.User.query.get(email) # or flask.abort(404)
    #  Workaround: if email doesn't exist, use relay
    if user is None:
        return flask.jsonify("smtp:[{}]:{}".format("smtp-relay.gmail.com", "25"))
    #  End workaround
    return flask.jsonify(user.email)


@internal.route("/postfix/transport/<path:email>")
def postfix_transport(email):
    if email == '*' or re.match("(^|.*@)\[.*\]$", email):
        return flask.abort(404)
    #  Workaround: if email doesn't exist, use relay
    if models.User.query.get(email) is None:
        return flask.jsonify("smtp:[{}]:{}".format("smtp-relay.gmail.com", "25"))
    #  End workaround
    localpart, domain_name = models.Email.resolve_domain(email)
    relay = models.Relay.query.get(domain_name) or flask.abort(404)
    return flask.jsonify("smtp:[{}]".format(relay.smtp))


@internal.route("/postfix/sender/login/<path:sender>")
def postfix_sender_login(sender):
    localpart, domain_name = models.Email.resolve_domain(sender)
    if localpart is None:
        return flask.abort(404)
    destination = models.Email.resolve_destination(localpart, domain_name, True)
    # return flask.jsonify(",".join(destination)) if destination else flask.abort(404)
    # Workaround: allow anybody from our local domain
    if destination:
        return flask.jsonify(",".join(destination))
    elif models.Domain.query.get(domain_name) is not None:
        return flask.jsonify("{}@{}".format(localpart, domain_name))
    else:
        return flask.abort(404)
    # End workaround


@internal.route("/postfix/sender/access/<path:sender>")
def postfix_sender_access(sender):
    """ Simply reject any sender that pretends to be from a local domain
    """
    if not is_void_address(sender):
        localpart, domain_name = models.Email.resolve_domain(sender)
        # Workaround: allow anybody from our local domain
        if models.Domain.query.get(domain_name) is None:
            return flask.jsonify("REJECT")
        else:
            return flask.jsonify("OK")
        # End workaround
        # return flask.jsonify("REJECT") if models.Domain.query.get(domain_name) else flask.abort(404)
    else:
        return flask.abort(404)
```

**Comment by Nebukadneza** [comment]
Hi There,

The `Mailu`-Project is currently in a bit of a bind! We are short on man-power, and we need to judge if it is possible for us to put in some work on this issue.

To help with that, we are currently trying to find out which issues are actively keeping users from using `Mailu`, which issues have someone who want to work on them — and which issues may be less important. These a less important ones could be discarded for the time being, until the project is in a more stable and regular state once again.

In order for us to better assess this, it would be helpful if you could put a **reaction on this post** (use the :smiley: icon to the top-right).
- 👍️ if you **need this** to be able to use Mailu. Ideally, you’d also be able to test this on your installation, and provide feedback …
- 🎉 if you find it a **nice bonus**, but no deal-breaker
- 🚀 if you want to **work on it yourself**!
We want to keep this **voting open for 2 weeks** from now, so please help out!

**Comment by deantheiceman** [comment]
I'm using the docker images, is there any quick fix I can use to implement this? @deze333 

Thanks

**Comment by deze333** [comment]
@deantheiceman I've done it manually by editing `postfix.py` file as described above, plus some small tweak in `main.cf`. That seemed to work but I can't guarantee it'll work for every situation.

**Comment by Diman0** [comment]
Hi There,

we see this issue had only little attention 😟. As much as it pains us: In order to get the current issues a bit more manageable for us, we decided to close it. 👋👋

We hope that no one feels offended by doing so. Should the issue get really pressing in the future, please feel free to re-open it.

Thank you for your patience and understanding, 🙇
- Your Mailu Team

**Comment by HollyIT** [comment]
Just ran into this same problem today, and figure a bunch more people might be looking for a solution with Google forcing people off the free GSuite to the paid Workspaces. Here's a workaround I came up with.

So say you want me@example.com to be handled by Google.

**First on MailU:**

Create a new account, say me_relay@example.com. Next setup an alias for me@example.com to me_relay@example.com.

**Next in GMail (from the account you're keeping on GSuite):**

From GMail go to settings and under accounts add a new account under "Check mail from other accounts". Have it check the account you created above (me_relay@example.com).

It's not the prettiest thing, but it is working.
